File: implementations/python/provekit_lsp/forward_propagator.py
"""ForwardPropagator — accumulate posts and emit implication-check diagnostics.
Per: docs/lsp/forward-propagation-floor-v1.md
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class Post:
    constraints: list[str]
    is_top: bool

    @staticmethod
    def top() -> "Post":
        return Post(constraints=[], is_top=True)

    @staticmethod
    def of(constraint: str) -> "Post":
        return Post(constraints=[constraint], is_top=False)


@dataclass
class DiagnosticResult:
    code: str
    message: str


class ForwardPropagator:
    def __init__(self):
        self._seed_catalog: dict[str, Post] = {}

    def add_to_catalog(self, callee_id: str, pre: Post, post: Post) -> None:
        self._seed_catalog[callee_id] = post

    def check_callsite(
        self, callee_id: str, current_post: Post
    ) -> Optional[DiagnosticResult]:
        if current_post.is_top:
            return None
        callee_pre = self._seed_catalog.get(callee_id)
        if callee_pre is None:
            return None
        for c in current_post.constraints:
            if c not in callee_pre.constraints:
                return DiagnosticResult(
                    code="implication-failed",
                    message=f"post does not imply callee pre: {' && '.join(callee_pre.constraints)}",
                )
        return None

File: implementations/rust/provekit-lsp/src/forward_propagator.rs
/// ForwardPropagator — accumulate posts and emit implication-check diagnostics.
/// Per: docs/lsp/forward-propagation-floor-v1.md

#[derive(Debug, Clone)]
pub struct Post {
    pub constraints: Vec<String>,
    pub is_top: bool,
}

impl Post {
    pub fn top() -> Self {
        Self { constraints: vec![], is_top: true }
    }

    pub fn of(constraint: impl Into<String>) -> Self {
        Self { constraints: vec![constraint.into()], is_top: false }
    }
}

#[derive(Debug, Clone)]
pub struct DiagnosticResult {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Default)]
pub struct ForwardPropagator {
    seed_catalog: std::collections::HashMap<String, Post>,
}

impl ForwardPropagator {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_to_catalog(&mut self, callee_id: &str, _pre: Post, post: Post) {
        self.seed_catalog.insert(callee_id.to_string(), post);
    }

    pub fn check_callsite(&self, callee_id: &str, current_post: &Post) -> Option<DiagnosticResult> {
        if current_post.is_top {
            return None;
        }
        let callee_pre = self.seed_catalog.get(callee_id)?;
        for constraint in &current_post.constraints {
            if !callee_pre.constraints.contains(constraint) {
                return Some(DiagnosticResult {
                    code: "implication-failed".to_string(),
                    message: format!(
                        "post does not imply callee pre: {}",
                        callee_pre.constraints.join(" && ")
                    ),
                });
            }
        }
        None
    }
}